chain/x/shard/keeper/keeper_test.go

```diff
@@ -41,137 +43,117 @@ func (s *TestSuite) SetupTest() {
-		},
-	}
-	for _, batch := range batches {
-		_, err := s.keeper.SubmitBatch(s.ctx, &types.SubmitBatchRequest{
-			Sender:           s.auth,
-			TransactionBatch: batch,
-		})
-		s.Require().NoError(err)
+func (s *TestSuite) TestExportGenesis() {
+	submit1 := &types.SubmitCardinalTxRequest{
+		Sender:    s.auth,
+		Namespace: "foo",
+		Tick:      1,
+		Txs: &types.Transactions{Txs: []*types.Transaction{
+			{1, []byte("foo")},
+			{10, []byte("bar")},
+			{1, []byte("baz")},
+		}},
 	}
 
-	// submit one not relevant to our namespace.
-	_, err := s.keeper.SubmitBatch(s.ctx, &types.SubmitBatchRequest{
-		Sender: s.auth,
-		TransactionBatch: &types.TransactionBatch{
-			Namespace: "notcardinal",
-			Tick:      32,
-			Batch:     []byte("foo"),
-		},
-	})
-	s.Require().NoError(err)
+	submit2 := &types.SubmitCardinalTxRequest{
+		Sender:    s.auth,
+		Namespace: "bar",
+		Tick:      3,
+		Txs: &types.Transactions{Txs: []*types.Transaction{
+			{15, []byte("qux")},
+			{2, []byte("quiz")},
+		}},
+	}
 
-	res, err := s.keeper.Batches(s.ctx, &types.QueryBatchesRequest{
-		Namespace: ns,
-		Page:      nil,
-	})
-	s.Require().NoError(err)
-	s.Require().Len(res.Batches, len(batches))
-
-	// limit the request to only 2.
-	limit := uint32(2)
-	res, err = s.keeper.Batches(s.ctx, &types.QueryBatchesRequest{
-		Namespace: ns,
-		Page: &types.PageRequest{
-			Key:   nil,
-			Limit: limit,
-		},
-	})
-	s.Require().NoError(err)
-	// should only have received 2.
-	s.Require().Len(res.Batches, int(limit))
-
-	// query again with the key in page response should give us the remaining batch.
-	res, err = s.keeper.Batches(s.ctx, &types.QueryBatchesRequest{
-		Namespace: ns,
-		Page: &types.PageRequest{
-			Key: res.Page.Key,
-		},
-	})
-	s.Require().NoError(err)
-	s.Require().Len(res.Batches, len(batches)-int(limit))
-}
+	submit3 := &types.SubmitCardinalTxRequest{
+		Sender:    s.auth,
+		Namespace: "foo",
+		Tick:      2,
+		Txs: &types.Transactions{Txs: []*types.Transaction{
+			{4, []byte("qux")},
+			{9, []byte("quiz")},
+		}},
+	}
 
-// TestSubmitBatch_DuplicateTick tests that when duplicate ticks are submitted, the data is overwritten.
-func (s *TestSuite) TestSubmitBatch_DuplicateTick() {
-	batch := &types.TransactionBatch{
-		Namespace: "cardinal",
-		Tick:      4,
-		Batch:     []byte("data"),
+	reqs := []*types.SubmitCardinalTxRequest{submit1, submit2, submit3}
+	for _, req := range reqs {
+		_, err := s.keeper.SubmitCardinalTx(s.ctx, req)
+		s.Require().NoError(err)
 	}
 
-	_, err := s.keeper.SubmitBatch(s.ctx, &types.SubmitBatchRequest{
-		Sender:           s.auth,
-		TransactionBatch: batch,
-	})
-	s.Require().NoError(err)
+	gen := s.keeper.ExportGenesis(s.ctx)
+	// there should only be 2 namespaced txs, because we only submitted 2 diff ones.
+	s.Require().Len(gen.Txs, 2)
+	s.Require().Len(gen.Txs[1].Txs, 2)            // we submitted 2 ticks for namespace "foo"
+	s.Require().Len(gen.Txs[1].Txs[0].Txs.Txs, 3) // the first tick had 3 txs
```

Comment:
Holy cow, this is confusing. I think these "Txs" should be renamed to better reflect what they actually are. e.g.: `gen.Namespace[1].Tick[0].Txs`. Can the second field in the TickedTransaction struct be a `repeated Transaction` instead of a struct that only contains the repeated transaction?

Reply:
okay simplified this a bunch - the storage of the transactions were wrapped so i could store them like this:

`store(namespace) -> key = tick number : value = transactions`

wrapping the transactions in their own proto message allows you to proto.Marshal them entirely self contained (i can't proto.Marshal a slice of proto messages). but, i don't think this really gives us any real benefit, and ended up just being confusing.

i switched it to just marshal the tick. this does add redundancy, so now you have 

`store(namespace) -> key = tick number : value = types.Tick(tick, []txs)`

so tick number gets stored twice, but i dont think it should be much of a problem

Reply:
👍